packages/core/src/preagg/PreAggregator.js

```diff
@@ -331,20 +331,45 @@ function preaggregateInfo(clientQuery, active, preaggCols, schema) {
 
 /**
  * Push column selections down to subqueries.
+ * @param {Query} query The (sub)query to push down to.
+ * @param {string[]} cols The column names to push down.
  */
 function subqueryPushdown(query, cols) {
   const memo = new Set;
   const pushdown = q => {
+    // it is possible to have duplicate subqueries
+    // so we memoize and exit early if already seen
     if (memo.has(q)) return;
     memo.add(q);
+
     if (isSelectQuery(q) && q._from.length) {
+      // select the pushed down columns
+      // note that the select method will deduplicate for us
       q.select(cols);
+      if (isAggregateQuery(q)) {
+        // if an aggregation query, we need to push to groupby as well
+        // we also deduplicate as the column may already be present
+        const set = new Set(
+          q._groupby.flatMap(x => x instanceof ColumnNameRefNode ? x.name : []
+        ));
+        q.groupby(cols.filter(c => !set.has(c)));
```

Comment:
Wouldn't it be easier to deduplicate in the method for adding clauses/column refs to the groupby?

Reply:
I looked into that first, but its a bit trickier to handle the general case where all kinds of different expressions get passed in. For example, one could first render everything out to strings and then dedupe based on those. Here it seemed simpler: we only need to check against raw column references as we know the inputs are limited to string column names.